### PR TITLE
Add WICD unit test job

### PR DESCRIPTION
**Pull request by sebsoto**
This PR adds a new job to run against WMCO PRs.
This job spins up a Windows VM, and runs the WICD unit test make target.

This job has been set as not required, as the unit test make target does not yet exist.

### CODE REVIEW

File: ci-operator/step-registry/windows/unit/wicd/windows-unit-wicd-ref.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+ref:
+  as: windows-unit-wicd
+  from: src
+  commands: windows-unit-wicd-commands.sh
+  resources:
+    requests:
+      cpu: 1000m
+      memory: 400Mi
+  timeout: 2h15m0s
```
Comment (saifshaikh48) [inline ci-operator/step-registry/windows/unit/wicd/windows-unit-wicd-ref.yaml:9]: I expected this test to be much shorter than our other e2e jobs. Any reason for the long timeout?

File: ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh [hunk 2]
```diff
@@ -37,6 +49,9 @@ vm_template_username=$(<"${SHARED_DIR}"/windows_vm_template_username.txt)
 # set VM folder
 vm_folder="/${vsphere_datacenter}/vm/${cluster_name}"
```
Comment (jrvaldes) [inline ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh:50]: Given no cluster is being created, it may be misleading to lean on `${cluster_name}`?

File: ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh [hunk 3]
```diff
@@ -17,6 +17,18 @@ function createWindowsInstanceFile() {
       echo "$(date -u --rfc-3339=seconds) - created ${windows_instance_file}"
 }
 
+# Creates a folder at the given path if it does not exist.
+# Will return error if the parent folder also does not exist.
+# Takes the folder's path as an argument.
+function ensureSubfolderExists() {
```
Comment (jrvaldes) [inline ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh:23]: s/ensureSubfolderExists()/ensureVMFolderExists()

File: ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh [hunk 4]
```diff
@@ -17,6 +17,18 @@ function createWindowsInstanceFile() {
...
+      folder=$1;
```
Comment (jrvaldes) [inline ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh:24]: s/folder/PATH

File: ci-operator/step-registry/upi/vsphere/windows/post/upi-vsphere-windows-post-commands.sh [hunk 5]
```diff
@@ -23,3 +23,12 @@ do
         echo "$(date -u --rfc-3339=seconds) - $(rm -vf "${f}")"
     fi
 done
+
+# remove any created folders. Each line in the created_folders.txt file will contain the name of a folder
+folder_file="${SHARED_DIR}/created_folders.txt"
+while read -r line; do
+    echo "$(date -u --rfc-3339=seconds) - destroying folder ${line}..."
```
Comment (jrvaldes) [inline ci-operator/step-registry/upi/vsphere/windows/post/upi-vsphere-windows-post-commands.sh:30]:
```suggestion
    echo "$(date -u --rfc-3339=seconds) - removing folder ${line}"
```

File: ci-operator/step-registry/upi/vsphere/windows/post/upi-vsphere-windows-post-commands.sh [hunk 6]
```diff
@@ -23,3 +23,12 @@ do
...
+    govc object.destroy "${line}"
+    # cleanup
+done < $folder_file
```
Comment (jrvaldes) [inline ci-operator/step-registry/upi/vsphere/windows/post/upi-vsphere-windows-post-commands.sh:33]: This may fail if `$folder_file` does not exist.

File: ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh [hunk 7]
```diff
@@ -17,6 +17,18 @@ function createWindowsInstanceFile() {
...
+      if ! govc folder.info $folder; then
+          echo "$(date -u --rfc-3339=seconds) - creating folder $folder..."
+          govc folder.create $folder
+	  echo "${folder}" >> "${SHARED_DIR}/created_folders.txt"
```
Comment (jrvaldes) [inline ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh:28]: `${SHARED_DIR}` is shared across all steps and the file can be overwritten, resulting in a resource leak. I suggest naming this file `windows_vm_folders.txt` to lower the collision rate with other files.

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 8]
```diff
@@ -146,6 +146,17 @@ tests:
...
     - ref: windows-e2e-operator-test-byoh
     workflow: windows-e2e-operator-test-platform-none-vsphere
+- as: wicd-unit-vsphere
+  steps:
+    cluster_profile: vsphere-platform-none
+    post:
+    - chain: upi-vsphere-windows-post
+    pre:
+    - ref: ipi-install-rbac
+    - chain: upi-conf-vsphere-platform-none-ovn
```
Comment (jrvaldes) [inline ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:156]: This chain includes unnecessary steps, for example, `ovn-conf-hybrid-manifest-with-custom-vxlan-port`.

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 9]
```diff
@@ -146,6 +146,17 @@ tests:
...
+    - chain: upi-vsphere-windows-pre
```
Comment (jrvaldes) [inline ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:157]: I feel it should be beneficial to create an explicit workflow just with above `pre` and `post` chains, where this test may read:

```yaml
- as: wicd-unit-vsphere
  steps:
    cluster_profile: vsphere-platform-none
    test:
    - ref: windows-unit-wicd
    workflow: upi-vsphere-windows-install
```

Also, it can be reused for other tests.

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml [hunk 10]
```diff
@@ -146,6 +146,17 @@ tests:
     test:
     - ref: windows-e2e-operator-test-byoh
     workflow: windows-e2e-operator-test-platform-none-vsphere
+- as: wicd-unit-vsphere
+  steps:
+    cluster_profile: vsphere-platform-none
```
Comment (jrvaldes) [inline ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:151]: I suggest using `cluster_profile: vsphere`.

**Review by jrvaldes — COMMENTED** [review]
@sebsoto  Thanks for working on this, comments added. PTAL

**Comment by sebsoto** [comment]
@alinaryan we can see from the logs that the two main goals of this PR are being accomplished
```
 INFO[2022-03-29T19:52:41Z] 2022-03-29 19:52:30+00:00 - Processing Windows instance file: /tmp/secret/192.168.56.18_windows_instance.txt
make: *** No rule to make target `wicd-unit'.  Stop. 
```
A Windows instance is created, and the `wicd-unit` make target is being called.
These are the requirements to test changes being made in the WMCO repo.

File: ci-operator/step-registry/windows/unit/wicd/windows-unit-wicd-ref.yaml (hunk 1 above)
Comment (jrvaldes) [reply ci-operator/step-registry/windows/unit/wicd/windows-unit-wicd-ref.yaml:9]: You should lower the `requests` for resources too.

File: ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh (hunk 7 above)
Comment (selansen) [reply ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh:28]: +1 for above comment

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 10 above)
Comment (selansen) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:151]: any reason for choosing platform-none?

Comment (sebsoto) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:151]: I don't believe its possible to use `cluster_profile: vsphere`

Comment (sebsoto) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:151]: > any reason for choosing platform-none?

Yes because it allows for us to spin up resources and have network access to them without creating an OpenShift cluster

File: ci-operator/step-registry/upi/vsphere/windows/post/upi-vsphere-windows-post-commands.sh (hunk 5 above)
Comment (sebsoto) [reply ci-operator/step-registry/upi/vsphere/windows/post/upi-vsphere-windows-post-commands.sh]: I'd like to stick to destroying as that is the language used by the govc command.

File: ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh (hunk 4 above)
Comment (sebsoto) [reply ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh:24]: I dont want to collide with the existing PATH variable

File: ci-operator/step-registry/windows/unit/wicd/windows-unit-wicd-commands.sh [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+# parse instance's address from filename
+INSTANCE_ADDRESS=$(basename "${instance_file}" "_windows_instance.txt")
+export INSTANCE_ADDRESS
+# grab user to SSH as from the file's content
+INSTANCE_USERNAME=$(cat $instance_file)
+export INSTANCE_USERNAME
+# export key required to SSH into the instance
+export KUBE_SSH_KEY_PATH=${CLUSTER_PROFILE_DIR}/ssh-privatekey
+
+make wicd-unit
```
Comment (rvanderp3) [inline ci-operator/step-registry/windows/unit/wicd/windows-unit-wicd-commands.sh:29]: This is failing in pj-rehearse.  Is the intent to add required make targets outside the scope of this PR?

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 8 above)
Comment (rvanderp3) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:156]: There are a number of unnecessary steps being pulled in to this workflow.  For example, step `upi-conf-vsphere-ova` ensures that the required OVA for the installer associated with the CI job exists if it doesn't the OVA gets uploaded.  I don't see anything really problematic since the likelihood of the OVA not already existing is somewhat low.   It looks like steps could be omitted from the workflow:
- wicd-unit-vsphere-ovn-conf
- wicd-unit-vsphere-ovn-conf-hybrid-manifest-with-custom-vxlan-port
- wicd-unit-vsphere-upi-conf-vsphere-platform-none
- wicd-unit-vsphere-upi-conf-vsphere
- wicd-unit-vsphere-upi-conf-vsphere-ova

It might be a good idea to create a new chain which only pulls in the needed steps.

**Review by rvanderp3 — COMMENTED** [review]
@sebsoto overall looks good, I left a couple of comments.

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 10 above)
Comment (jrvaldes) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:151]: AFAIK, the `cluster_profile` just provides the `LEASED_RESOURCE`.  The proposed VMs can be created in any ci-segment, I don't see any constraints.

In addition, using `vsphere` as cluster_profile will alleviate the leasing resources for `platform-none` jobs, where there are only 4 ci-segments.

File: ci-operator/step-registry/windows/unit/wicd/windows-unit-wicd-commands.sh (hunk 11 above)
Comment (sebsoto) [reply ci-operator/step-registry/windows/unit/wicd/windows-unit-wicd-commands.sh:29]: It is, the make target has now been added, and it is passing

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 10 above)
Comment (sebsoto) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml:151]: > Is this job able to run in either?

I don't see why not, I might be missing something here though

>  How many concurrent instances of this job do you think will be running?

Hopefully not too many :)
This should be a quicker job, but it will run against all WMCO PRs

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 9 above)
Comment (sebsoto) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml]: i'd prefer to hold off on creating a workflow for this until there is another case where we need a windows VM without a cluster
this seems like a one-off, and I'd like to not clutter the step-registry further, without good reason of course

File: ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml (hunk 8 above)
Comment (sebsoto) [reply ci-operator/config/openshift/windows-machine-config-operator/openshift-windows-machine-config-operator-master.yaml]: I've cut down on the steps but would personally like to avoid creating a new chain, [see here for reasoning](https://github.com/openshift/release/pull/27251#discussion_r844355278)

File: ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh (hunk 2 above)
Comment (sebsoto) [reply ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh:50]: Considering this will mainly be used when a cluster is created I'd like to just leave it as is

File: ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh [hunk 12]
```diff
@@ -13,10 +13,22 @@ function createWindowsInstanceFile() {
       # set file name
       windows_instance_file="${SHARED_DIR}/${ADDRESS}_windows_instance.txt"
       # create file with instance information
-      echo "username=${USERNAME}" > "${windows_instance_file}"
```
Comment (jrvaldes) [inline ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh:16]: Why this change?

Comment (jrvaldes) [reply ci-operator/step-registry/upi/vsphere/windows/pre/upi-vsphere-windows-pre-commands.sh:16]: The username is already available in `${SHARED_DIR}"/windows_vm_template_username.txt`.

On the other hand, `${ADDRESS}_windows_instance.txt` serves as the raw content of the instance information, allowing it to be re-used and extended. Adding just the `username` to the content of the file takes that flexibility away.

If the intention is to maintain a file with just the username of the Windows instance, you should consider something like: `${ADDRESS}_windows_instance_username.txt`.

**Review by jrvaldes — APPROVED** [review]
/lgtm

Nice work @sebsoto, thanks for working on this.
